belongs_to with ParameterizedType

elielhaouzi opened the pull request:
#3434 

makes belongs_to function accepts options for ParameterizedType other that the default ones.

I assume that once the options passed to the belongs_to fields are valid, the field is right initiated as we see in the `check_field_type!` function.

josevalim commented:
Good work! As mentioned in #3434, I would still be careful about mixing both but given the PR effectively allows them to be used together, we can at least explore how it goes and maybe revert it if indeed there complications and the whole pattern should be discouraged. :) Thanks! :green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

elielhaouzi commented:
My bad, maybe the ParameterizedPrefixedString module that I added to types.exs may be confusing/wrong.

[There](https://github.com/elixir-ecto/ecto/blob/master/integration_test/support/types.exs#L36), the dump function returns the result of the cast function as though the cast need to return the Ecto native type but it is not true, the cast function need to return the runtime data as mentioned in the Ecto.Type doc.

I did the same error in [ParameterizedTypePrefixedUUID](https://github.com/elixir-ecto/ecto/blob/master/test/ecto/repo/autogenerate_test.exs#L74) and in [ParameterizedTypePrefixedID](https://github.com/elixir-ecto/ecto/blob/master/test/ecto/repo/autogenerate_test.exs#L92) 

This is the fix for the ParameterizedPrefixedString:
```elixir
defmodule ParameterizedPrefixedString do
  use Ecto.ParameterizedType
  def init(opts), do: Enum.into(opts, %{})
  def type(_), do: :string

  def cast(value, %{prefix: prefix}) do
    if String.starts_with?(value, [prefix <> "-"] do
      {:ok, value}
    else
      {:ok, prefix <> "-" <> value}
    end
  end

  def dump(string, _, %{prefix: _prefix} = opts), do: {:ok, string |> String.split("-") |> List.last()}
  def load(string, _, %{prefix: prefix}), do: {:ok, prefix <> "-" <> string}
  def embed_as(_, _), do: :dump
end
```
it is not critical because in all cases it returns a string but can conduct to error to future dev that see the code.

To do an another PR to fix this?

josevalim commented:
Good find. Yes, please do send a PR! :heart:
